### ISSUE TITLE
Add `selector-shadow-no-unmatchable`

### ISSUE THREAD

**Issue by renet** [stylelint/stylelint#5575]
> What is the problem you're trying to solve?

My team develops a web components library, which comes with the usage of css selectors like `:host` and `::slotted()`. Especially the last one allows only for selectors on the 1st level (light DOM "skin"), as explained [here](https://stackoverflow.com/a/61631668/1014782). Thus, selecting any child of a slotted element is not possible, like for example:

```css
::slotted(p span) { /*...*/ }

/* or */

::slotted(p) span { /* ... */ }
```

for a DOM like

```html
<custom-component>
  <p>Hello <span>world</span>.</p>
</custom-component>
```

But it's easy, especially for devs new to web components / the shadow DOM, to make that mistake and try to use selectors like these. Right now, there seems to be no rule that prevents a developer from making these mistakes.

> What solution would you like to see?

A new rule that throws an error, if a developer tries to write a selector, that by definition can never select anything. Afaik, this would at least apply to selectors like:

```css
::slotted(div a) {}
::slotted(div) a {}
::slotted(::slotted(div)) {}
::slotted(a):hover {}
::slotted(> div) {}
::slotted(:host) {}
::slotted(div::before) {}
div :host {}
:host(> .class) {}
:host(div a) {}

/* but also probably selectors like */

div::after a {}
div::after::before {}

/* and probably many more. */
```

If there's rules for what I am looking for, then I'm sorry, but I didn't find any.  I assume, that the [selector-disallowed-list](https://stylelint.io/user-guide/rules/list/selector-disallowed-list/) rule might be used as a workaround to achieve what I want, but this rule is actually part of the set of rules used to "limit language features" and I expect the above examples to be explicitly not part of the language features. Thanks for any replies in advance!

**Comment by Mouvedia** [comment]
If we were to add a rule what about encompassing all functions that require one "simple selector"?
One potential option of such a rule would be something like `not-lvl-3` (boolean) which would reject `:not(div, .foo)` and require the more cross-browser `:not(div):not(.foo)`.

ref https://caniuse.com/css-not-sel-list

**Comment by jeddy3** [comment]
With the possibilities in https://github.com/stylelint/stylelint/issues/5907 and @Mouvedia's suggested feature above, it feels like there's lots of scope for linting selectors that are unmatchable in some way.

Let's break these into small and more focused rules.

Let's limit this issue to disallowing unmatchable shadow dom selectors.

- Name: `selector-shadow-no-unmatchable`
- Primary option: `true`
- Autofixable: No
- Message: `Unexpected unmatchable shadow selector "$selectors"`
- Description: "Disallow unmatchable shadow selectors"
- Extended description: (need something here about how they aren't matchable)
- Section: "Avoid errors" -> "Selectors"

I've labelled the issue as ready to implement. Please consider [contributing](https://stylelint.io/contributing) if you have time.

There are [steps on how to add a new rule](https://stylelint.io/developer-guide/rules#add-a-rule) in the Developer guide.

@Mouvedia Can you [open a new issue](https://github.com/stylelint/stylelint/issues/new?assignees=&labels=&template=REQUEST_A_FEATURE.yml) for your suggestion so that we can flesh out the details?

**Comment by Mouvedia** [comment]
> Can you [open a new issue](https://github.com/stylelint/stylelint/issues/new?assignees=&labels=&template=REQUEST_A_FEATURE.yml) for your suggestion so that we can flesh out the details?

This is too specific (niche cross-browser + only one pseudo-class) to warrant its own rule.
That's why I was advocating for a single rule that cover all cases* in which the function requires a "simple selector".

<sub>\* to be tallied, 2 for now</sub>

**Comment by jeddy3** [comment]
> That's why I was advocating for a single rule that cover all cases* in which the function requires a "simple selector".

Yes, that sounds like a good scope for the rule.  Let's open a new issue so that we flesh it out.

**Comment by Mouvedia** [comment]
> > That's why I was advocating for a single rule that cover all cases* in which the function requires a "simple selector".
> 
> Yes, that sounds like a good scope for the rule. Let's open a new issue so that we flesh it out.

#5974
